### ISSUE TITLE
Validate no-Mosaic phase boundary under positive-worker prefetch

### ISSUE THREAD

**Issue by shapovalovei** [shapovalovei/YOLOX-OBB#75]
## Phase

Bounded framework validation only. No implementation, training, optimizer step, GPU work, dataset mutation, export, or CI.

## Origin

Downstream `card-detector-training` issue #64 audited exact epoch-boundary resume semantics for production candidate #43 and exposed a likely generic YOLOX trainer/DataLoader phase-boundary defect.

Downstream reference:

https://github.com/shapovalovei/card-detector-training/issues/64

This issue exists to validate the generic framework behavior before any downstream workaround is considered.

## Baselines

Downstream pinned framework:

- commit `0695e2834ff5716dfd51fcd6ac7b2300b7bc27ce`
- tree `22b2aaa3b4d018a60bb7363f15fa5838b6c5f07c`

Current YOLOX-OBB `main` at issue creation:

- commit `219bc22e9362954e5da0beef182e7b276116d2d9`

First verify whether the behavior is present on both the pinned downstream commit and current `main`.

## Suspected defect

The maintained trainer creates one long-lived `DataPrefetcher` before the epoch loop.

With positive DataLoader workers, PyTorch prefetches future batch-index tasks. The maintained `YoloBatchSampler` embeds the current `mosaic` boolean in each emitted tuple:

`(input_dim, idx, mosaic)`

`Dataset.resize_getitem` applies that tuple value when the worker later executes the sample.

The trainer calls:

`train_loader.close_mosaic()`

inside `before_epoch()` at the no-aug boundary.

But changing `batch_sampler.mosaic` only affects **future** tuples. Any already-prefetched tuples retain the old `mosaic=True` value.

For the downstream #43 runtime contract:

- workers = 4
- default PyTorch `prefetch_factor=2`
- one DataPrefetcher-held next batch

there can be one held batch plus eight queued DataLoader tasks already emitted before `close_mosaic()` runs.

Hypothesis: the first batches of the intended no-Mosaic epoch can still execute Mosaic even though `close_mosaic()` has been called before that epoch's training loop.

This would make the maintained phase boundary semantically stale under positive-worker prefetch.

## Goal

Prove or falsify whether `close_mosaic()` at an epoch boundary guarantees that the **very first optimizer-consumed batch after the boundary** is no-Mosaic when positive workers/prefetch are enabled.

This must be tested for both HBB and OBB maintained dataset wrappers where practical because the tuple/decorator mechanism is shared.

## Read first

- `AGENTS.md`
- current/open framework issues touching Mosaic/DataLoader/prefetch, including #53/#65 for adjacent runtime-state evidence
- `yolox/core/trainer.py`
- `yolox/data/data_prefetcher.py`
- `yolox/data/dataloading.py`
- `yolox/data/samplers.py`
- `yolox/data/datasets/datasets_wrapper.py`
- HBB `MosaicDetection`
- OBB `MosaicDetectionOBB`
- PyTorch DataLoader prefetch behavior for the validated runtime

## Required validation

Use deterministic synthetic data only.

No model or optimizer is required.

### A. Zero-worker control

With `num_workers=0`, establish expected behavior:

1. consume an enabled-Mosaic batch stream;
2. call `close_mosaic()` at a controlled logical boundary;
3. verify the next consumed batch has `mosaic=False`.

### B. Positive-worker reproduction

Use at minimum:

- `num_workers=4`
- `prefetch_factor=2`

Construct the actual maintained:

`InfiniteSampler/YoloBatchSampler -> DataLoader -> DataPrefetcher` lifecycle where CUDA is not required, or an equivalent CPU boundary that preserves DataLoader task prefetch semantics.

If `DataPrefetcher` itself cannot run CPU-only, instrument the underlying DataLoader iterator while preserving the same one-held-next-batch logic explicitly and state the limitation.

Before `close_mosaic()` record:

- logical batches already consumed;
- batch tasks already emitted by the sampler;
- embedded mosaic flags for emitted future tasks where observable;
- number of outstanding prefetched tasks.

Then call `close_mosaic()` and inspect the next consumed batches.

Report exactly how many post-boundary batches still carry `mosaic=True`.

### C. Actual tuple semantics

Prove that the stale behavior, if present, is caused by already-emitted `(input_dim, idx, mosaic=True)` tuples rather than a worker-local stale dataset boolean.

Show that newly emitted tuples after `close_mosaic()` carry `mosaic=False`.

### D. HBB / OBB

Cover the shared decorated `__getitem__` path for HBB and OBB sufficiently to prove whether both inherit the same phase-boundary behavior.

Do not infer one from the other without checking the maintained wrappers.

### E. Pinned vs current main

Return the verdict separately for:

- downstream pinned `0695e283...`
- current `main` at validation time

If current main already fixes it, identify the exact fixing commit and stop.

## Correctness contract

For an exact no-aug boundary, after the declared transition point:

- first consumed batch must use `mosaic=False`;
- no queued stale `mosaic=True` tuple may cross the boundary;
- L1/no-aug phase bookkeeping must correspond to the same first batch.

Do not reinterpret a handful of stale Mosaic batches as acceptable just because the epoch-level flag changed.

## Decision

Return exactly one:

A — **NO DEFECT**

Positive-worker prefetch does not leak Mosaic batches across the phase boundary.

B — **GENERIC PREFETCH PHASE-BOUNDARY DEFECT PROVEN**

Already-emitted Mosaic tuples cross `close_mosaic()` and the first no-aug epoch is not exact.

C — **PINNED-ONLY DEFECT**

The downstream pinned commit is affected but current main already has a verified generic fix.

D — **INCONCLUSIVE**

The maintained path cannot be reproduced faithfully enough.

## If B

Do NOT implement in this issue.

Return the smallest generic correctness boundary for a later implementation issue. Evaluate at least:

- safely resetting/reconstructing DataLoader/prefetch state at `close_mosaic()`;
- a phase-generation/token approach that invalidates stale prefetched tasks;
- another framework-owned mechanism.

Preserve data order/RNG semantics as much as possible and explicitly state any effect on reproducibility.

Do not recommend a card-detector-training-only patch for a generic defect.

## Scope boundaries

NO TRAINING.
NO OPTIMIZER STEP.
NO GPU / CUDA session.
NO real dataset.
NO model-quality evaluation.
NO tracked file modification.
NO branch/commit/PR.
NO export.
NO CI.

Temporary probes stay outside the repository.

## Handoff

Post a factual handoff containing:

- exact refs validated;
- worker/prefetch configuration;
- zero-worker control;
- positive-worker emitted/consumed mosaic flags;
- count of stale post-boundary batches;
- HBB result;
- OBB result;
- root cause;
- FACT / INFERENCE / UNVERIFIED;
- A/B/C/D verdict;
- smallest framework-owned follow-up if defect proven.

STOP for orchestrator review. Do not close this issue yourself.

**Comment by shapovalovei** [comment]
## Research handoff — Issue #75

Validation-only work is complete. No tracked repository files, branches, commits, PRs, CI, model, optimizer, training, GPU/CUDA, real data, export, or model-quality work was used. The YOLOX-OBB worktree remains clean.

### Identity and source facts

- Repository: `shapovalovei/YOLOX-OBB`.
- Refreshed `origin/main` with `git fetch origin main`; current `HEAD == origin/main == 219bc22e9362954e5da0beef182e7b276116d2d9`.
- Pinned validation ref: `0695e2834ff5716dfd51fcd6ac7b2300b7bc27ce`.
- Local validation runtime: Python `3.11.11`, PyTorch `2.2.2`, CUDA unavailable.
- The Issue #64 latest orchestrator comment routes this generic trainer/DataLoader boundary to YOLOX-OBB before accepting the downstream resume design.
- At both refs, `trainer.py` constructs one `DataPrefetcher` before the epoch loop, and `before_epoch()` calls `train_loader.close_mosaic()` at the no-aug boundary.
- `DataPrefetcher` holds one result from `next(loader)` at construction and calls `preload()` immediately after every `next()` return.
- `YoloBatchSampler.__iter__()` emits `(input_dim, idx, self.mosaic)` tuples. `DataLoader.close_mosaic()` only assigns `batch_sampler.mosaic = False`.
- `Dataset.resize_getitem()` assigns the tuple's dimension and mosaic flag to the worker-local wrapper before invoking the decorated `__getitem__`.
- The relevant `DataPrefetcher`, `dataloading`, `samplers`, and `datasets_wrapper` blobs are identical at the pinned and current refs. Current main's separate `3b3c334` change only corrects the enabled-Mosaic runtime-dimension source in the HBB/OBB wrappers; it does not address phase/prefetch state.

The real `DataPrefetcher` was not instantiated because its maintained implementation unconditionally creates `torch.cuda.Stream()` and calls `.cuda()`. The probe therefore used a CPU-only logical equivalent: request one held batch at construction; each probe `next()` returns that held batch and immediately requests the following DataLoader batch. This preserves the relevant DataLoader task/prefetch lifecycle without invoking CUDA.

### Probe

The temporary probe used only synthetic data: 16 synthetic items, batch size 2, input dimension `(8, 10)`, `InfiniteSampler(shuffle=True, seed=0)`, maintained `YoloBatchSampler`, maintained `DataLoader`, and the maintained HBB or OBB decorated wrapper. Worker Python/NumPy/Torch seeds were initialized deterministically. Probe SHA-256: `eeb83eb089fd1be9b76d50aa8c01c48585ba6a3548f13a625b53c7188f8fd0b1`.

The HBB and OBB probe outputs independently matched the embedded tuple flag: Mosaic samples reached the preprocessor with the Mosaic output shape and no-Mosaic samples reached it with the raw synthetic image shape. Thus the result covers the maintained decorated `__getitem__` path, not only detached sampler output.

### Results

Each A2 run finished the eight logical batches in the synthetic `DataLoader` before calling `close_mosaic()`.

| control | HBB pinned | OBB pinned | HBB current | OBB current |
|---|---:|---:|---:|---:|
| A1 raw DataLoader: first post-close batch | `False` | `False` | `False` | `False` |
| A2 zero-worker held batch: stale `True` count | 1 | 1 | 1 | 1 |
| A2 positive-worker stale `True` count | 9 | 9 | 9 | 9 |

For every positive-worker run (`num_workers=4`, `prefetch_factor=2`), immediately before `close_mosaic()`:

- 8 logical batches had been consumed by the CPU DataPrefetcher equivalent;
- 17 sampler tasks had been emitted in total: 8 consumed tasks, held task ordinal 8, and eight future DataLoader task records ordinals 9–16;
- every emitted tuple before the boundary had `mosaic=True`;
- `_task_info` contained exactly eight outstanding future task records, keys 9–16;
- PyTorch's `_tasks_outstanding` counter varied with whether workers had already placed results in the result queue (8 on the current HBB/OBB runs, 6 on the pinned HBB/OBB runs), but `_task_info` still contained the same eight not-yet-consumed task records.

Post-boundary consumption was identical for HBB and OBB, and for both refs:

```text
ordinal 8   embedded=True   observed=True    held-by-prefetcher
ordinal 9   embedded=True   observed=True    already-emitted DataLoader task
ordinal 10  embedded=True   observed=True    already-emitted DataLoader task
ordinal 11  embedded=True   observed=True    already-emitted DataLoader task
ordinal 12  embedded=True   observed=True    already-emitted DataLoader task
ordinal 13  embedded=True   observed=True    already-emitted DataLoader task
ordinal 14  embedded=True   observed=True    already-emitted DataLoader task
ordinal 15  embedded=True   observed=True    already-emitted DataLoader task
ordinal 16  embedded=True   observed=True    already-emitted DataLoader task
ordinal 17  embedded=False  observed=False   newly-emitted after close_mosaic()
```

Therefore the exact stale count is 9: 1 held-by-prefetcher plus 8 already-emitted DataLoader tasks. No newly emitted post-close task was stale; the first newly emitted tuple and the first newly consumed newly-emitted batch were `mosaic=False`.

The A2 zero-worker result is also material: after the eight logical batches, the held ordinal-8 batch was `True`; `close_mosaic()` then caused the next newly emitted ordinal-9 tuple and consumed batch to be `False`. Thus the DataPrefetcher-held batch alone crosses the boundary even with no worker queue. Positive workers increase the stale depth by eight.

A1 used no held-batch wrapper. It consumed ordinal 0 with `True`, called `close_mosaic()`, then requested ordinal 1; the newly emitted and consumed batch was `False` for both wrappers and both refs.

### Causality

**FACT.** Already-emitted tuples retain their original boolean because the Boolean is materialized by `YoloBatchSampler` at yield time. `close_mosaic()` mutates only the sampler's future `mosaic` value; it cannot mutate tuples already held by `DataPrefetcher` or already sent to workers/result queues.

**FACT.** Every tuple emitted after `close_mosaic()` in the probe had `mosaic=False`, and the worker-observed HBB/OBB decorated paths matched that tuple value. This rules out a persistent worker-local Mosaic Boolean as the cause.

**INFERENCE.** The first intended no-Mosaic optimizer batch is not exact under the maintained long-lived positive-worker lifecycle. The same phase violation exists without workers because the long-lived DataPrefetcher has one held batch.

### Smallest generic follow-up boundaries if implementation is later authorized

No fix was implemented. These are bounded framework-owned options for a later implementation decision:

| option | sample ordering | worker / augmentation RNG | reproducibility | no-aug exactness |
|---|---|---|---|---|
| A. Rebuild/reset the phase-scoped DataLoader and DataPrefetcher at the boundary | A naive reconstruction restarts `InfiniteSampler` and changes/duplicates order. Preserve order only by transferring a sampler cursor or deterministically fast-forwarding to the exact next logical task. | Worker processes restart and normal worker seeds restart/change. Discarded prefetched work may already have advanced Python/NumPy/Torch RNG in the old workers; exact continuation requires a defined seed/cursor replay or explicit RNG state protocol. | Deterministic only if sampler position, loader seed, worker assignment, and augmentation RNG replay are part of the contract; reset alone is recipe-equivalent, not trajectory-identical. | Yes: construct the replacement loader with `mosaic=False` before its first preload. This is the smallest clear boundary for exact phase semantics, but it must not silently weaken order/RNG guarantees. |
| B. Add phase-generation/token invalidation for stale work | Exact ordering requires rejected stale tasks to be reissued at the same logical positions. Dropping them advances or skips the stream; ordinary DataLoader workers have no generic requeue operation. | A worker that executes rejected work can advance its RNG before invalidation. Exactness needs rewind/reseed, non-executing validation before augmentation, or stateless per-sample augmentation. | A token alone is insufficient; queue timing, reissue order, worker RNG, and augmentation state must also be controlled. This is materially more invasive than a phase-scoped rebuild. | Possible only with a framework-owned consumer/worker cutover that guarantees no stale result is returned and replacement work is false; not provided by the current DataLoader contract. |
| C. Drain/fence the old iterator before changing phase | Consuming the held/queued Mosaic batches preserves the current stream but moves the phase boundary by up to 9 batches. Discarding them changes order. Neither matches the declared epoch-85/86 boundary. | Consuming preserves those RNG advances but applies the wrong augmentation phase; discarding loses their RNG history and requires replay to preserve exact continuation. | Can be deterministic, but only by accepting a shifted boundary or a new discard/replay contract. | No for the declared boundary; it makes the no-aug transition late. |

Preferred later direction: an explicit phase-scoped loader/prefetcher handoff owned by YOLOX-OBB, with sampler position and RNG/replay semantics specified separately. Do not use a card-detector-training-only workaround for this generic defect. A later implementation issue must decide whether “exact” means only the first post-boundary batch is no-Mosaic or also bitwise-identical sample/augmentation continuation.

### Verdict

- `0695e2834ff5716dfd51fcd6ac7b2300b7bc27ce`: **B — GENERIC PREFETCH PHASE-BOUNDARY DEFECT PROVEN**.
- `219bc22e9362954e5da0beef182e7b276116d2d9` (refreshed `origin/main`): **B — GENERIC PREFETCH PHASE-BOUNDARY DEFECT PROVEN**.

The current ref does not contain a generic fix. STOP for orchestrator review; no implementation is authorized by this validation.

**Comment by shapovalovei** [comment]
## Orchestrator acceptance

Accepted as **B — GENERIC PREFETCH PHASE-BOUNDARY DEFECT PROVEN** on both validated refs:

- pinned downstream framework `0695e2834ff5716dfd51fcd6ac7b2300b7bc27ce`;
- current `main` `219bc22e9362954e5da0beef182e7b276116d2d9`.

Independent review confirms the causal chain is sufficiently isolated:

- raw zero-worker control transitions immediately for newly requested work;
- trainer-equivalent one-held prefetch crosses exactly one stale Mosaic batch even with zero workers;
- `num_workers=4`, `prefetch_factor=2` adds eight already-emitted DataLoader task records, yielding exactly nine stale post-boundary Mosaic batches;
- HBB and OBB decorated `__getitem__` paths both reproduce the same behavior;
- all tuples emitted after `close_mosaic()` carry `mosaic=False`, ruling out persistent worker-local Mosaic state;
- the defect therefore belongs to the generic trainer/DataLoader/prefetch phase boundary, not downstream card-detector-training.

The CPU logical prefetch wrapper is acceptable evidence for this validation because the maintained CUDA `DataPrefetcher` control-flow property under test is exactly one held `next(loader)` result plus immediate preload after every returned batch; no CUDA-specific scheduling is required to establish tuple immutability or DataLoader task emission.

**Implementation is not yet authorized.** The handoff correctly exposes a real design choice that must be decided before coding:

1. phase-boundary exactness only — guarantee the first post-transition optimizer batch is no-Mosaic, while permitting a changed sample/augmentation trajectory at the cutover; or
2. trajectory-preserving cutover — guarantee no stale Mosaic crosses the boundary while also preserving the uninterrupted logical sample order and augmentation RNG semantics.

A naive loader/prefetch rebuild satisfies (1) but can restart/shift sampler and worker RNG state. A phase token alone does not satisfy (2) unless rejected work is reissued with deterministic sampler/RNG replay. This is therefore a design/contract decision, not something to hide inside a minimal patch.

No downstream workaround is accepted. Any implementation must remain framework-owned in YOLOX-OBB after owner/orchestrator chooses the required preservation contract.

Validation complete; closing #75 as Completed.
